sale_promotion_programs/models/sale_coupon_program.py

```diff
@@ -0,0 +1,116 @@
+        elif not self._filter_on_mimimum_amount(order):
+            message = {
+                "error": _("A minimum of %s %s should be purchased to get the reward")
+                % (self.rule_minimum_amount, self.currency_id.name)
+            }
+        elif self.promo_code and self.promo_code == order.promo_code:
+            message = {"error": _("The promo code is already applied on this order")}
+        elif not self.promo_code and self in order.no_code_promo_program_ids:
+            message = {
+                "error": _("The promotional offer is already applied on this order")
+            }
+        elif not self.active:
+            message = {"error": _("Promo code is invalid")}
+        elif (
+            self.rule_date_from
+            and self.rule_date_from > order.date_order
+            or self.rule_date_to
+            and order.date_order > self.rule_date_to
+        ):
+            message = {"error": _("Promo code is expired")}
+        elif order.promo_code and self.promo_code_usage == "code_needed":
+            message = {"error": _("Promotionals codes are not cumulative.")}
+        elif (
+            self._is_global_discount_program()
+            and order._is_global_discount_already_applied()
+        ):
+            message = {"error": _("Global discounts are not cumulative.")}
+        elif (
+            self.promo_applicability == "on_current_order"
+            and self.reward_type == "product"
+            and not order._is_reward_in_order_lines(self)
+        ):
+            message = {
+                "error": _(
+                    "The reward products should be in the sales order lines to \
+                     apply the discount."
+                )
+            }
+        elif not self._is_valid_partner(order.partner_id):
+            message = {"error": _("The customer doesn't have access to this reward.")}
+        elif not self._filter_programs_on_products(order):
+            message = {
+                "error": _(
+                    "You don't have the required product quantities on your \
+                    sales order.If the reward is same product quantity, please \
+                     make sure that all the products are recorded on the sales \
+                      order (Example: You need to have 3 T-shirts on your sales \
+                      order if the promotion is 'Buy 2, Get 1 Free'."
+                )
+            }
+        else:
+            if (
+                self not in applicable_programs
+                and self.promo_applicability == "on_current_order"
+            ):
+                message = {
+                    "error": _(
+                        "At least one of the required conditions is not met \
+                        to get the reward!"
+                    )
+                }
+        return message
+
+    def _filter_programs_by_sale_order_amount(self, order):
```

Comment:
```suggestion
    def _filter_programs_by_sale_order_count(self, order):
```

language issue: reading sale_order_amount, people will thing of the total price of the sale order, not the number of sale orders.